File: examples/typescript/consumer/src/scenarios.ts
import { getClient, getQueuePrefix } from "./config.js";
import { log } from "./logger.js";

export interface ScenarioStep {
  action: string;
  detail: string;
}

export interface ScenarioRunResult {
  queueId: string;
  steps: ScenarioStep[];
}

export type ScenarioName = "basic" | "ack-deadletter" | "priority" | "sessions" | "idempotency";

export interface ScenarioDescriptor {
  name: ScenarioName;
  role: "consumer";
  description: string;
  run: () => Promise<ScenarioRunResult>;
}

const EMPTY_QUEUE_DETAIL = "queue empty — run the producer's scenario first";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** basic - dequeues with ack (count=5, default ttl) and immediately acknowledges every item returned. */
async function runBasic(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-basic`;
  const client = getClient();
  const steps: ScenarioStep[] = [];

  const result = await client.dequeueLocked(queueId, { count: 5 });
  if (!result || result.items.length === 0) {
    log("WARN", EMPTY_QUEUE_DETAIL);
    steps.push({ action: "dequeue", detail: EMPTY_QUEUE_DETAIL });
    return { queueId, steps };
  }

  const dequeueDetail = `Dequeued ${result.items.length} item(s) from queue ${queueId}`;
  log("INFO", dequeueDetail);
  steps.push({ action: "dequeue", detail: dequeueDetail });

  for (const dequeued of result.items) {
    await client.acknowledge(queueId, dequeued.lockId);
    const detail = `Acknowledged item ${JSON.stringify(dequeued.item)} (lockId=${dequeued.lockId})`;
    log("INFO", detail);
    steps.push({ action: "acknowledge", detail });
  }

  return { queueId, steps };
}

/**
 * ack-deadletter - dequeues 3 items (ttl=10s) and branches on each item's `outcome` field: acks
 * "ack", dead-letters "deadletter" (logging the resulting DLQ id), and leaves "expire" locked.
 * Then waits for the lock to expire, re-dequeues to show redelivery, and finally drains the DLQ.
 */
async function runAckDeadletter(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-ackdlq`;
  const dlqId = `${queueId}-deadletter`;
  const client = getClient();
  const steps: ScenarioStep[] = [];

  const result = await client.dequeueLocked(queueId, { count: 3, ttlSeconds: 10 });
  if (!result || result.items.length === 0) {
    log("WARN", EMPTY_QUEUE_DETAIL);
    steps.push({ action: "dequeue", detail: EMPTY_QUEUE_DETAIL });
    return { queueId, steps };
  }

  const lockIds = result.items.map((i) => i.lockId);
  const dequeueDetail = `dequeued ${result.items.length} items, lockIds=[${lockIds.join(", ")}]`;
  log("INFO", dequeueDetail);
  steps.push({ action: "dequeue", detail: dequeueDetail });

  for (const dequeued of result.items) {
    const payload = dequeued.item as { outcome?: string; n?: number };
    const outcome = payload.outcome ?? "ack";

    if (outcome === "deadletter") {
      await client.deadLetter(queueId, dequeued.lockId);
      const detail = `dead-lettered item outcome=deadletter -> dlqId=${dlqId}`;
      log("INFO", detail);
      steps.push({ action: "deadletter", detail });
    } else if (outcome === "expire") {
      const detail = `leaving item outcome=expire locked to expire (n=${payload.n})`;
      log("INFO", detail);
      steps.push({ action: "skip", detail });
    } else {
      await client.acknowledge(queueId, dequeued.lockId);
      const detail = `acked item outcome=${outcome} (n=${payload.n})`;
      log("INFO", detail);
      steps.push({ action: "acknowledge", detail });
    }
  }

  const waitDetail = "waiting 11s for outcome=expire item's lock to expire";
  log("INFO", waitDetail);
  steps.push({ action: "wait", detail: waitDetail });
  await sleep(11000);

  const redelivered = await client.dequeueLocked(queueId, { count: 3, ttlSeconds: 10 });
  if (redelivered && redelivered.items.length > 0) {
    const redeliverDetail = `dequeued ${redelivered.items.length} item(s) again via redelivery`;
    log("INFO", redeliverDetail);
    steps.push({ action: "dequeue", detail: redeliverDetail });

    for (const dequeued of redelivered.items) {
      await client.acknowledge(queueId, dequeued.lockId);
      const detail = `acknowledged redelivered item ${JSON.stringify(dequeued.item)}`;
      log("INFO", detail);
      steps.push({ action: "acknowledge", detail });
    }
  } else {
    const detail = "no redelivered item found on second dequeue";
    log("WARN", detail);
    steps.push({ action: "dequeue", detail });
  }

  const dlqResult = await client.dequeueLocked(dlqId, { count: 1 });
  if (dlqResult && dlqResult.items.length > 0) {
    const dlqItem = dlqResult.items[0];
    const dequeueDlqDetail = `dequeued dead-lettered item from ${dlqId}: ${JSON.stringify(dlqItem.item)}`;
    log("INFO", dequeueDlqDetail);
    steps.push({ action: "dequeue", detail: dequeueDlqDetail });

    await client.acknowledge(dlqId, dlqItem.lockId);
    const ackDlqDetail = `acknowledged item in ${dlqId} (drained)`;
    log("INFO", ackDlqDetail);
    steps.push({ action: "acknowledge", detail: ackDlqDetail });
  } else {
    const detail = `${dlqId} empty — no dead-lettered item to drain`;
    log("WARN", detail);
    steps.push({ action: "dequeue", detail });
  }

  return { queueId, steps };
}

/**
 * priority - dequeues with ack (count=10) in one call and acknowledges each item as it's returned,
 * demonstrating that the fast-lane (priority=0) items surface before the normal ones despite
 * being enqueued later.
 */
async function runPriority(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-priority`;
  const client = getClient();
  const steps: ScenarioStep[] = [];

  const result = await client.dequeueLocked(queueId, { count: 10 });
  if (!result || result.items.length === 0) {
    log("WARN", EMPTY_QUEUE_DETAIL);
    steps.push({ action: "dequeue", detail: EMPTY_QUEUE_DETAIL });
    return { queueId, steps };
  }

  const order = result.items.map((i) => (i.item as { n?: number }).n);
  const dequeueDetail = `dequeued ${result.items.length} item(s) in order n=[${order.join(", ")}]`;
  log("INFO", dequeueDetail);
  steps.push({ action: "dequeue", detail: dequeueDetail });

  for (const dequeued of result.items) {
    await client.acknowledge(queueId, dequeued.lockId);
    const n = (dequeued.item as { n?: number }).n;
    const detail = `acknowledged item n=${n} (priority=${dequeued.priority})`;
    log("INFO", detail);
    steps.push({ action: "acknowledge", detail });
  }

  return { queueId, steps };
}

const KNOWN_SESSION_IDS = ["session-a", "session-b"];

async function drainSession(
  queueId: string,
  sessionId: string,
  leaseId: string,
  steps: ScenarioStep[],
): Promise<void> {
  const client = getClient();
  const sessionQueueId = `${queueId}-session-${sessionId}`;
  const dequeued = await client.dequeueLocked(sessionQueueId, { count: 10, leaseId });

  if (!dequeued || dequeued.items.length === 0) {
    const detail = `session queue ${sessionQueueId} empty`;
    log("WARN", detail);
    steps.push({ action: "dequeue", detail });
  } else {
    const order = dequeued.items.map((i) => (i.item as { seq?: number }).seq);
    const dequeueDetail = `dequeued ${dequeued.items.length} item(s) from ${sessionQueueId} in order seq=[${order.join(", ")}]`;
    log("INFO", dequeueDetail);
    steps.push({ action: "dequeue", detail: dequeueDetail });

    for (const item of dequeued.items) {
      await client.acknowledge(sessionQueueId, item.lockId, { leaseId });
      const seq = (item.item as { seq?: number }).seq;
      const detail = `acknowledged item seq=${seq} for session=${sessionId}`;
      log("INFO", detail);
      steps.push({ action: "acknowledge", detail });
    }
  }

  await client.releaseSession(queueId, sessionId, leaseId);
  const releaseDetail = `released session=${sessionId} leaseId=${leaseId}`;
  log("INFO", releaseDetail);
  steps.push({ action: "release-session", detail: releaseDetail });
}

/**
 * sessions - round 1 accepts "any available" session (server picks whichever of the known
 * sessions is unclaimed); round 2 targets whichever known session round 1 didn't return (both,
 * if round 1 found nothing available). Each round dequeues with ack against the derived session
 * queue id, acknowledges each item, then releases the session.
 */
async function runSessions(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-sessions`;
  const client = getClient();
  const steps: ScenarioStep[] = [];

  // Round 1: "any available" mode (no sessionId) - the server picks whichever of the
  // known sessions is currently unclaimed.
  let claimedFirst: string | undefined;
  const anyLease = await client.acceptSession(queueId, { leaseSeconds: 30 });
  if (!anyLease) {
    const detail = "no session currently available (any-available mode)";
    log("WARN", detail);
    steps.push({ action: "accept-session", detail });
  } else {
    claimedFirst = anyLease.sessionId;
    const detail = `accepted session=${claimedFirst} via any-available mode leaseId=${anyLease.leaseId}`;
    log("INFO", detail);
    steps.push({ action: "accept-session", detail });
    await drainSession(queueId, claimedFirst, anyLease.leaseId, steps);
  }

  // Round 2: targeted mode - accept whichever known session round 1 didn't return
  // (both, if round 1 found nothing available).
  for (const sessionId of KNOWN_SESSION_IDS) {
    if (sessionId === claimedFirst) {
      continue;
    }

    const lease = await client.acceptSession(queueId, { sessionId, leaseSeconds: 30 });
    if (!lease) {
      const detail = `session ${sessionId} not available (targeted mode) — run the producer's scenario first`;
      log("WARN", detail);
      steps.push({ action: "accept-session", detail });
      continue;
    }

    const acceptDetail = `accepted session=${sessionId} via targeted mode leaseId=${lease.leaseId}`;
    log("INFO", acceptDetail);
    steps.push({ action: "accept-session", detail: acceptDetail });
    await drainSession(queueId, sessionId, lease.leaseId, steps);
  }

  return { queueId, steps };
}

/**
 * idempotency - dequeues from the same queue the producer's idempotency scenario just filled.
 * Expects exactly 2 items (n=1 and n=3) since n=2, the duplicate, should never have been
 * enqueued at all - logs that expectation explicitly and acknowledges whatever is returned.
 */
async function runIdempotency(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-idempotency`;
  const client = getClient();
  const steps: ScenarioStep[] = [];

  const result = await client.dequeueLocked(queueId, { count: 5 });
  if (!result || result.items.length === 0) {
    log("WARN", EMPTY_QUEUE_DETAIL);
    steps.push({ action: "dequeue", detail: EMPTY_QUEUE_DETAIL });
    return { queueId, steps };
  }

  const order = result.items.map((i) => (i.item as { n?: number }).n);
  const dequeueDetail = `dequeued ${result.items.length} item(s) n=[${order.join(", ")}] - expecting 2 (n=1, n=3); the duplicate n=2 should have been silently dropped by idempotency dedup`;
  log("INFO", dequeueDetail);
  steps.push({ action: "dequeue", detail: dequeueDetail });

  for (const dequeued of result.items) {
    await client.acknowledge(queueId, dequeued.lockId);
    const n = (dequeued.item as { n?: number }).n;
    const detail = `acknowledged item n=${n}`;
    log("INFO", detail);
    steps.push({ action: "acknowledge", detail });
  }

  return { queueId, steps };
}

export const SCENARIOS: ScenarioDescriptor[] = [
  { name: "basic", role: "consumer", description: "Dequeues items with ack and acknowledges each one.", run: runBasic },
  {
    name: "ack-deadletter",
    role: "consumer",
    description: "Acks, dead-letters, and lets one item expire, then drains the DLQ.",
    run: runAckDeadletter,
  },
  {
    name: "priority",
    role: "consumer",
    description: "Dequeues items to demonstrate fast-lane items surfacing before normal ones.",
    run: runPriority,
  },
  {
    name: "sessions",
    role: "consumer",
    description: "Accepts each session in turn, dequeues and acknowledges its items, then releases it.",
    run: runSessions,
  },
  {
    name: "idempotency",
    role: "consumer",
    description: "Dequeues and acknowledges the survivors of the producer's dedup demo.",
    run: runIdempotency,
  },
];

File: examples/python/consumer/scenarios.py
"""Consumer-side implementation of the 4 demo scenarios.

See examples/shared/SCENARIOS.md for the exact business logic each of these
functions must follow. Each function dequeues/acks/dead-letters items on
`queue_id` using `client` (a `daprmq_client.DaprMQClient`), logs every step
via `log(level, message)`, and returns the list of `{"action", "detail"}`
step dicts for the `POST /scenarios/{name}/run` response body.

**Empty-queue tolerance**: per API_CONTRACT.md, a consumer scenario finding
nothing to dequeue (e.g. the producer's scenario hasn't run yet) is not an
error — it returns 200 with a step noting the empty queue.
"""
from __future__ import annotations

import asyncio
from collections.abc import Callable

from daprmq_client import DaprMQClient

LogFn = Callable[[str, str], None]

EMPTY_QUEUE_DETAIL = "queue empty — run the producer's scenario first"


async def run_basic(client: DaprMQClient, queue_id: str, log: LogFn) -> list[dict]:
    result = await client.dequeue_locked(queue_id, count=5)
    if result is None or not result.items:
        log("INFO", f"Queue {queue_id} empty — run the producer's scenario first")
        return [{"action": "dequeue", "detail": EMPTY_QUEUE_DETAIL}]

    log("INFO", f"Dequeued {len(result.items)} items from {queue_id}")
    steps = [{"action": "dequeue", "detail": f"dequeued {len(result.items)} items"}]

    for it in result.items:
        await client.acknowledge(queue_id, it.lock_id)
        log("INFO", f"Acknowledged item {it.item} (lockId={it.lock_id})")
        steps.append({"action": "acknowledge", "detail": f"acked item {it.item}"})

    return steps


async def run_ack_deadletter(client: DaprMQClient, queue_id: str, log: LogFn) -> list[dict]:
    dlq_id = f"{queue_id}-deadletter"

    result = await client.dequeue_locked(queue_id, count=3, ttl_seconds=10)
    if result is None or not result.items:
        log("INFO", f"Queue {queue_id} empty — run the producer's scenario first")
        return [{"action": "dequeue", "detail": EMPTY_QUEUE_DETAIL}]

    lock_ids = [it.lock_id for it in result.items]
    log("INFO", f"Dequeued {len(result.items)} items from {queue_id}, lockIds={lock_ids}")
    steps = [{"action": "dequeue", "detail": f"dequeued {len(result.items)} items, lockIds={lock_ids}"}]

    expire_pending = False
    for it in result.items:
        outcome = (it.item or {}).get("outcome", "ack") if isinstance(it.item, dict) else "ack"

        if outcome == "deadletter":
            await client.dead_letter(queue_id, it.lock_id)
            log("INFO", f"Dead-lettered item outcome=deadletter -> dlqId={dlq_id}")
            steps.append({"action": "deadletter", "detail": f"dead-lettered item outcome=deadletter -> dlqId={dlq_id}"})
        elif outcome == "expire":
            expire_pending = True
            log("INFO", "Leaving item outcome=expire locked so its lock can expire")
        else:
            await client.acknowledge(queue_id, it.lock_id)
            log("INFO", f"Acknowledged item outcome={outcome}")
            steps.append({"action": "acknowledge", "detail": f"acked item outcome={outcome}"})

    if expire_pending:
        log("INFO", "Waiting 11s for outcome=expire item's lock to expire")
        steps.append({"action": "wait", "detail": "waiting 11s for outcome=expire item's lock to expire"})
        await asyncio.sleep(11)

        redelivered = await client.dequeue_locked(queue_id, count=3, ttl_seconds=10)
        if redelivered is not None and redelivered.items:
            log("INFO", f"Dequeued {len(redelivered.items)} redelivered item(s) from {queue_id}")
            steps.append({"action": "dequeue", "detail": "dequeued expired item again via redelivery"})
            for it in redelivered.items:
                await client.acknowledge(queue_id, it.lock_id)
                log("INFO", f"Acknowledged redelivered item outcome=expire ({it.item})")
                steps.append({"action": "acknowledge", "detail": "acked redelivered item outcome=expire (drained)"})
        else:
            log("WARN", f"Expected a redelivered item from {queue_id} but none arrived")
            steps.append({"action": "dequeue", "detail": "no redelivered item found on retry (unexpected)"})

    dlq_result = await client.dequeue_locked(dlq_id, count=1, ttl_seconds=10)
    if dlq_result is not None and dlq_result.items:
        for it in dlq_result.items:
            log("INFO", f"Dequeued dead-lettered item from {dlq_id}: {it.item}")
            steps.append({"action": "dequeue", "detail": f"dequeued dead-lettered item from {dlq_id}: {it.item}"})
            await client.acknowledge(dlq_id, it.lock_id)
            log("INFO", f"Acknowledged item from {dlq_id} (drained)")
            steps.append({"action": "acknowledge", "detail": f"acked item from {dlq_id} (drained)"})
    else:
        log("INFO", f"DLQ {dlq_id} empty — nothing to drain")
        steps.append({"action": "dequeue", "detail": f"DLQ {dlq_id} empty — nothing to drain"})

    return steps


async def run_priority(client: DaprMQClient, queue_id: str, log: LogFn) -> list[dict]:
    result = await client.dequeue_locked(queue_id, count=10)
    if result is None or not result.items:
        log("INFO", f"Queue {queue_id} empty — run the producer's scenario first")
        return [{"action": "dequeue", "detail": EMPTY_QUEUE_DETAIL}]

    order = [it.item.get("n") if isinstance(it.item, dict) else None for it in result.items]
    log("INFO", f"Dequeued {len(result.items)} items from {queue_id} in order n={order}")
    steps = [{"action": "dequeue", "detail": f"dequeued {len(result.items)} items in order n={order}"}]

    for it in result.items:
        await client.acknowledge(queue_id, it.lock_id)
        n = it.item.get("n") if isinstance(it.item, dict) else None
        log("INFO", f"Acknowledged item n={n} (priority={it.priority})")
        steps.append({"action": "acknowledge", "detail": f"acked item n={n} (priority={it.priority})"})

    return steps


KNOWN_SESSION_IDS = ("session-a", "session-b")


async def _drain_session(client: DaprMQClient, queue_id: str, session_id: str, lease_id: str, log: LogFn) -> list[dict]:
    steps: list[dict] = []
    session_queue_id = f"{queue_id}-session-{session_id}"
    try:
        dequeued = await client.dequeue_locked(session_queue_id, count=10, ttl_seconds=30, lease_id=lease_id)
        if dequeued is None or not dequeued.items:
            log("WARN", f"No items dequeued for session {session_id} from {session_queue_id}")
            steps.append(
                {"action": "dequeue", "detail": f"queue empty for session {session_id} — run the producer's scenario first"}
            )
        else:
            seqs = [it.item.get("seq") if isinstance(it.item, dict) else None for it in dequeued.items]
            log("INFO", f"Dequeued {len(dequeued.items)} items for session {session_id} in order seq={seqs}")
            steps.append(
                {
                    "action": "dequeue",
                    "detail": f"dequeued {len(dequeued.items)} items for session {session_id} in order seq={seqs}",
                }
            )
            for it in dequeued.items:
                await client.acknowledge(session_queue_id, it.lock_id, lease_id=lease_id)
                seq = it.item.get("seq") if isinstance(it.item, dict) else None
                log("INFO", f"Acknowledged item session={session_id} seq={seq}")
                steps.append({"action": "acknowledge", "detail": f"acked item session={session_id} seq={seq}"})
    finally:
        await client.release_session(queue_id, session_id, lease_id)
        log("INFO", f"Released session {session_id}")
        steps.append({"action": "release_session", "detail": f"released session {session_id}"})

    return steps


async def run_sessions(client: DaprMQClient, queue_id: str, log: LogFn) -> list[dict]:
    steps: list[dict] = []

    # Round 1: "any available" mode (session_id=None) - the server picks whichever
    # of the known sessions is currently unclaimed.
    claimed_first: str | None = None
    any_lease = await client.accept_session(queue_id, session_id=None, lease_seconds=30)
    if any_lease is None:
        log("WARN", f"No session currently available (any-available mode) on {queue_id}")
        steps.append({"action": "accept_session", "detail": "no session currently available (any-available mode)"})
    else:
        claimed_first = any_lease.session_id
        log("INFO", f"Accepted session {claimed_first} via any-available mode leaseId={any_lease.lease_id}")
        steps.append(
            {"action": "accept_session", "detail": f"accepted session {claimed_first} via any-available mode leaseId={any_lease.lease_id}"}
        )
        steps.extend(await _drain_session(client, queue_id, claimed_first, any_lease.lease_id, log))

    # Round 2: targeted mode - accept whichever known session round 1 didn't return
    # (both, if round 1 found nothing available).
    for session_id in KNOWN_SESSION_IDS:
        if session_id == claimed_first:
            continue

        lease = await client.accept_session(queue_id, session_id=session_id, lease_seconds=30)
        if lease is None:
            log("INFO", f"Session {session_id} not available (targeted mode) on {queue_id}")
            steps.append(
                {"action": "accept_session", "detail": f"session {session_id} not available (targeted mode)"}
            )
            continue

        log("INFO", f"Accepted session {session_id} via targeted mode leaseId={lease.lease_id}")
        steps.append(
            {"action": "accept_session", "detail": f"accepted session {session_id} via targeted mode leaseId={lease.lease_id}"}
        )
        steps.extend(await _drain_session(client, queue_id, session_id, lease.lease_id, log))

    return steps


async def run_idempotency(client: DaprMQClient, queue_id: str, log: LogFn) -> list[dict]:
    """Dequeues from the queue the producer's idempotency scenario just filled.
    Expects exactly 2 items (n=1 and n=3) since n=2, the duplicate, should never
    have been enqueued at all.
    """
    result = await client.dequeue_locked(queue_id, count=5)
    if result is None or not result.items:
        log("INFO", f"Queue {queue_id} empty — run the producer's scenario first")
        return [{"action": "dequeue", "detail": EMPTY_QUEUE_DETAIL}]

    order = [it.item.get("n") if isinstance(it.item, dict) else None for it in result.items]
    log(
        "INFO",
        f"Dequeued {len(result.items)} item(s) n={order} from {queue_id} - expecting 2 (n=1, n=3); "
        "the duplicate n=2 should have been silently dropped by idempotency dedup",
    )
    steps = [
        {
            "action": "dequeue",
            "detail": f"dequeued {len(result.items)} item(s) n={order} - expecting 2 (n=1, n=3)",
        }
    ]

    for it in result.items:
        await client.acknowledge(queue_id, it.lock_id)
        n = it.item.get("n") if isinstance(it.item, dict) else None
        log("INFO", f"Acknowledged item n={n}")
        steps.append({"action": "acknowledge", "detail": f"acked item n={n}"})

    return steps


RUNNERS: dict[str, Callable] = {
    "basic": run_basic,
    "ack-deadletter": run_ack_deadletter,
    "priority": run_priority,
    "sessions": run_sessions,
    "idempotency": run_idempotency,
}
